scalalib/src/mill/javalib/android/AndroidModule.scala

```diff
@@ -706,4 +708,248 @@ trait AndroidModule extends JavaModule {
+  }
+
+  def mainDexList: T[Option[PathRef]] = T {
+    None
+  }
+
+  def mainDexListOutput: T[Option[PathRef]] = T {
+    Some(PathRef(Task.dest / "runR8.dest/minify/main-dex-list-output.txt"))
+  }
+
+  // Optional baseline profile for ART rewriting
+  def baselineProfile: T[Option[PathRef]] = T {
+    None
+  }
+
+  def runR8: T[PathRef] = Task {
+
+    val destDir = Task.dest / "minify"
+    os.makeDir.all(destDir)
+
+    // Determine primary output:
+    // For apps, R8 will output DEX files into destDir (classes.dex, classes2.dex, ...).
+    // For libraries, R8 outputs a shrunk class jar.
+    val outputPath = if (isApplication()) destDir
+    else destDir / "classes-shrunk.jar"
+
+    T.log.debug("outptuPath: " + outputPath)
+
+    // Define diagnostic output file paths
+    val mappingOut = destDir / "mapping.txt"
+    val seedsOut = destDir / "seeds.txt"
+    val usageOut = destDir / "usage.txt"
+    val configOut = destDir / "configuration.txt"
+    destDir / "missing_rules.txt"
+    val baselineOutOpt = destDir / "baseline-profile-rewritten.txt"
+    destDir / "res"
+
+    // Create an extra ProGuard config file that instructs R8 to print seeds and usage.
+    val extraRulesFile = destDir / "extra-rules.pro"
+    val extraRulesContent =
+      s"""-printseeds ${seedsOut.toString}
+         |-printusage ${usageOut.toString}
+         |""".stripMargin.trim
+    os.write.over(extraRulesFile, extraRulesContent)
+
+    // Get the list of all class files to be processed by R8
+    super.compileClasspath().map(_.path).filter(os.isDir)
+      .flatMap(os.walk(_))
+      .filter(os.isFile)
+      .filter(_.ext == "class")
+      .map(_.toString())
+
+    val appCompiledFiles = os.walk(compile().classes.path)
+      .filter(_.ext == "class")
+      .map(_.toString)
+//        ++ inheritedClassFiles
+
+    T.log.debug(s"appCompiledFiles: ${appCompiledFiles}")
+
+    val r8ArgsBuilder = Seq.newBuilder[String]
+
+    r8ArgsBuilder += androidSdkModule().r8Path().path.toString
+
+    if (androidIsDebug()) {
+      r8ArgsBuilder += "--debug"
+    } else {
+      r8ArgsBuilder += "--release"
+    }
+
+    r8ArgsBuilder ++= Seq(
+      "--output",
+      outputPath.toString,
+      "--pg-map-output",
+      mappingOut.toString,
+      "--pg-conf-output",
+      configOut.toString
+    )
+
+    if (!enableDesugaring()) {
+      r8ArgsBuilder += "--no-desugaring"
+    }
+
+    if (isApplication()) {
+      r8ArgsBuilder ++= Seq(
+        "--min-api",
+        androidMinSdk().toString,
+        "--dex"
+      )
+    } else {
+      r8ArgsBuilder += "--classfile"
+    }
+
+    // Multi-dex arguments (if any are provided)
+    val multiDexArgs =
+      mainDexRules().toSeq.flatMap(r => Seq("--main-dex-rules", r.path.toString)) ++
+        mainDexList().toSeq.flatMap(l => Seq("--main-dex-list", l.path.toString)) ++
+        mainDexListOutput().toSeq.flatMap(l => Seq("--main-dex-list-output", l.path.toString))
```

Comment:
from the https://github.com/com-lihaoyi/mill/pull/4743/files#r2007060988 , you can decide the name of the output here and return it as part of a R8Output case class (needs a better name)